### PR TITLE
PP-8125 - New charges to use GatewayAccountCredential

### PR DESCRIPTION
**Pull request by nimalank7**
Description:
- New charges will use the payment_provider name from GatewayAccountCredential rather than GatewayAccount
- If there is only 1 GatewayAccountCredential then use that regardless of status otherwise get the latest ACTIVE GatewayAccountCredential
- Since there isn't a test for GatewayAccountEntity the logic had to be tested in ChargeServiceCreateTest. Since the logic of getGatewayName() is tested using charges only 1 new test
for telephone charges was added to verify that it correctly returned the right payment_provider name

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/gatewayaccount/model/GatewayAccountEntity.java [hunk 1]
```diff
@@ -180,7 +185,19 @@ public String getExternalId() {
     @JsonProperty("payment_provider")
     @JsonView(value = {Views.ApiView.class, Views.FrontendView.class})
     public String getGatewayName() {
-        return gatewayName;
+        List<GatewayAccountCredentialsEntity> gatewayAccountCredentialsEntities = getGatewayAccountCredentials();
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gatewayaccount/model/GatewayAccountEntity.java:189]: There are number of edge cases here. would be good to unit test this method.

File: src/test/java/uk/gov/pay/connector/charge/model/domain/ChargeEntityFixture.java [hunk 2]
```diff
@@ -66,10 +69,20 @@ public static GatewayAccountEntity defaultGatewayAccountEntity() {
         GatewayAccountEntity accountEntity = new GatewayAccountEntity("sandbox", new HashMap<>(), TEST);
         accountEntity.setId(1L);
         accountEntity.setServiceName("MyService");
+
         EmailNotificationEntity emailNotificationEntity = new EmailNotificationEntity(accountEntity);
         emailNotificationEntity.setTemplateBody("template body");
         accountEntity.addNotification(EmailNotificationType.PAYMENT_CONFIRMED, emailNotificationEntity);
         accountEntity.addNotification(EmailNotificationType.REFUND_ISSUED, emailNotificationEntity);
+
+        GatewayAccountCredentialsEntity gatewayAccountCredentialsEntity = new GatewayAccountCredentialsEntity(
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/charge/model/domain/ChargeEntityFixture.java]: We could use GatewayAccountCredentialsEntityFixture which comes with default and can be overridden

File: src/test/java/uk/gov/pay/connector/charge/service/Worldpay3dsFlexJwtServiceTest.java [hunk 3]
```diff
@@ -150,11 +157,19 @@ public void shouldCreateCorrectChallengeToken() {
         when(worldpay3dsFlexCredentialsEntity.getOrganisationalUnitId()).thenReturn("myOrg");
         when(worldpay3dsFlexCredentialsEntity.getJwtMacKey()).thenReturn("fa2daee2-1fbb-45ff-4444-52805d5cd9e0");
 
-        GatewayAccountEntity gatewayAccountEntity = aGatewayAccountEntity()
+        gatewayAccountCredentialsEntityList.add(new GatewayAccountCredentialsEntity(
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/charge/service/Worldpay3dsFlexJwtServiceTest.java]: Can we set default credentials list in `GatewayAccountEntityFixture` so we don't need to create add GatewayAccountCredentialsEntity everywhere. We can set the list explicitly if  `gatewayAccountCredentialsEntityList` need to be tested

File: src/main/java/uk/gov/pay/connector/gatewayaccount/model/GatewayAccountEntity.java (hunk 1 above)
Comment (nimalank7) [reply src/main/java/uk/gov/pay/connector/gatewayaccount/model/GatewayAccountEntity.java:189]: Done.

File: src/test/java/uk/gov/pay/connector/charge/model/domain/ChargeEntityFixture.java (hunk 2 above)
Comment (nimalank7) [reply src/test/java/uk/gov/pay/connector/charge/model/domain/ChargeEntityFixture.java]: Done.

File: src/test/java/uk/gov/pay/connector/charge/service/Worldpay3dsFlexJwtServiceTest.java (hunk 3 above)
Comment (nimalank7) [reply src/test/java/uk/gov/pay/connector/charge/service/Worldpay3dsFlexJwtServiceTest.java]: Done - in some cases `GatewayAccountCredentialsEntity` has been removed because it is already created as part of `GatewayAccountEntityFixture`.

File: src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java [hunk 4]
```diff
@@ -209,6 +212,18 @@ public void setUp() {
...
+                .aGatewayAccountCredentialsEntity()
+                .withGatewayAccountEntity(gatewayAccount)
+                .withPaymentProvider("sandbox")
+                .withCredentials(Map.of())
+                .withState(GatewayAccountCredentialState.CREATED)
+                .build();
+
+        List<GatewayAccountCredentialsEntity> gatewayAccountCredentialsEntities = new ArrayList<>();
+        gatewayAccountCredentialsEntities.add(gatewayAccountCredentialsEntity);
+        gatewayAccount.setGatewayAccountCredentials(gatewayAccountCredentialsEntities);
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java:225]: Could simplify
```suggestion
        gatewayAccount.setGatewayAccountCredentials(List.of(gatewayAccountCredentialsEntity));
```

File: src/test/java/uk/gov/pay/connector/gateway/processor/ChargeNotificationProcessorTest.java [hunk 5]
```diff
@@ -53,7 +48,9 @@
     
     @Before
     public void setUp() {
-        gatewayAccount = new GatewayAccountEntity("sandbox", new HashMap<>(), TEST);
+        gatewayAccount = GatewayAccountEntityFixture
+                .aGatewayAccountEntity()
+                .build();
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/gateway/processor/ChargeNotificationProcessorTest.java:53]: Could set ID on fixture
```suggestion
        gatewayAccount = GatewayAccountEntityFixture
                .aGatewayAccountEntity()
                .withId(GATEWAY_ACCOUNT_ID)
                .build();
```

File: src/test/java/uk/gov/pay/connector/service/RefundServiceTest.java [hunk 6]
```diff
@@ -410,14 +462,21 @@ public void shouldRefundAndSendEmail_whenGatewayRefundStateIsComplete_forChargeW
...
         String externalChargeId = "chargeId";
         Long accountId = 2L;
-        String providerName = "sandbox";
+        String providerName = SANDBOX.getName();
 
-        GatewayAccountEntity account = new GatewayAccountEntity(providerName, newHashMap(), TEST);
+        account = new GatewayAccountEntity(providerName, newHashMap(), TEST);
         account.setId(accountId);
+        account.setGatewayAccountCredentials(gatewayAccountCredentialsEntityList);
+        gatewayAccountCredentialsEntityList.add(GatewayAccountCredentialsEntityFixture
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/service/RefundServiceTest.java:470]: Possibly don't need this, if we change to use GatewayAccountEntityFixture

**Review by kbottla — APPROVED** [review]
LGTM! just a few very minor comments but not blocking
